### PR TITLE
[SPV-IR to OCL] Fix mutated builtin call attribute copying

### PR DESCRIPTION
**Pull request by wenju-he**
When SPIRVToOCL20Pass translates following SPV IR
```
  %call = tail call spir_func noundef ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4) noundef %Ptr, i32 noundef 0)
declare spir_func ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4), i32)
```
to OCL IR:
```
  %call = tail call spir_func noundef ptr @__to_private(ptr addrspace(4) noundef %Ptr)
declare spir_func ptr @__to_private(ptr addrspace(4))
```
, there is error `Attribute after last parameter` at copying attribute
to the new call because the old call has an additional AttributeSet for
the last argument. The last argument is eliminated in the new call.

Set tail call for new call if old call is tail call.

Revert cafd7e0. Partially revert c60327f.

### CODE REVIEW

File: lib/SPIRV/SPIRVBuiltinHelper.cpp [hunk 1]
```diff
@@ -103,7 +104,8 @@ Value *BuiltinCallMutator::doConversion() {
       Builder.Insert(addCallInst(CI->getModule(), FuncName, ReturnTy, Args,
                                  &Attrs, nullptr, Mangler.get()));
   NewCall->copyMetadata(*CI);
-  NewCall->setAttributes(CI->getAttributes());
+  NewCall->setAttributes(CallAttrs);
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVBuiltinHelper.cpp:107]: `setAttributes` + `getAttributes` also copies calling convention. Are we sure, that CallAttrs also contains it?

File: lib/SPIRV/SPIRVBuiltinHelper.cpp [hunk 2]
```diff
@@ -179,30 +185,21 @@ BuiltinCallMutator &BuiltinCallMutator::replaceArg(unsigned Index,
...
-      ArgAttrs.push_back(Attrs.getParamAttrs(I));
-    Attrs = AttributeList::get(CI->getContext(), Attrs.getFnAttrs(),
-                               Attrs.getRetAttrs(), ArgAttrs);
-  } else
-    moveAttributes(CI->getContext(), Attrs, Index + 1, Args.size() - Index - 1,
-                   Index);
+    Attrs = Attrs.removeParamAttributes(Ctx, Index);
+    CallAttrs = CallAttrs.removeParamAttributes(Ctx, Index);
+  } else {
+    moveAttributes(Ctx, Attrs, Index + 1, Args.size() - Index - 1, Index);
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVBuiltinHelper.cpp:200]: I'm not quite following, so call instruction and function declaration have different sets of attributes? Is it just about tail call?

**Review by MrSidims — COMMENTED** [review]
```
  %call = tail call spir_func noundef ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4) noundef %Ptr, i32 noundef 0)
declare spir_func ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4), i32)
to OCL IR:

  %call = tail call spir_func noundef ptr @__to_private(ptr addrspace(4) noundef %Ptr)
declare spir_func ptr @__to_private(ptr addrspace(4))
```
Can this be used as a test? You can add .spt file with usage of OpGenericCastToPtrExplicit instruction, translate it back like llvm-spirv -r --spirv-target-env="CL2.0"

File: lib/SPIRV/SPIRVBuiltinHelper.cpp (hunk 1 above)
Comment (wenju-he) [reply lib/SPIRV/SPIRVBuiltinHelper.cpp:107]: Calling convention isn't part of attributes. Calling convention for the new function is already set right after function creation at https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/1088aca0fd5ed0b855881bb52a199497dc86e952/lib/SPIRV/SPIRVUtil.cpp#L287

File: lib/SPIRV/SPIRVBuiltinHelper.cpp (hunk 2 above)
Comment (wenju-he) [reply lib/SPIRV/SPIRVBuiltinHelper.cpp:200]: Yes, they could be different.
As shown in the input IR in PR description:
The callInst has a `noundef` attribute for return value, and each argument also has `noundef` attribute. So the callinst has 3 AttributeSets.
On the other hand, the called function _Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi declaration doesn't have these 3 attributes.

Tail call isn't a part of callinst attributes. `NewCall->setTailCall(CI->isTailCall());` is added in this PR to ensure tail call isn't lost.

Original `Attrs` is only used for creating new function at https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/1088aca0fd5ed0b855881bb52a199497dc86e952/lib/SPIRV/SPIRVUtil.cpp#L289

**Comment by wenju-he** [comment]
> ```
>   %call = tail call spir_func noundef ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4) noundef %Ptr, i32 noundef 0)
> declare spir_func ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4), i32)
> to OCL IR:
> 
>   %call = tail call spir_func noundef ptr @__to_private(ptr addrspace(4) noundef %Ptr)
> declare spir_func ptr @__to_private(ptr addrspace(4))
> ```
> 
> Can this be used as a test? You can add .spt file with usage of OpGenericCastToPtrExplicit instruction, translate it back like llvm-spirv -r --spirv-target-env="CL2.0"

Can `noundef` be represented in .spt file? If not, callinst argument won't have it in spirv and the failure won't be triggered w/o this pr.
I didn't find an alterative attribute listed in https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_function_parameter_attribute that can be applied to the second argument of the call.

File: lib/SPIRV/SPIRVBuiltinHelper.cpp (hunk 1 above)
Comment (wenju-he) [reply lib/SPIRV/SPIRVBuiltinHelper.cpp:107]: The change at line 107 is the key change of this PR.

**Comment by wenju-he** [comment]
> ```
>   %call = tail call spir_func noundef ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4) noundef %Ptr, i32 noundef 0)
> declare spir_func ptr @_Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi(ptr addrspace(4), i32)
> to OCL IR:
> 
>   %call = tail call spir_func noundef ptr @__to_private(ptr addrspace(4) noundef %Ptr)
> declare spir_func ptr @__to_private(ptr addrspace(4))
> ```
> 
> Can this be used as a test? You can add .spt file with usage of OpGenericCastToPtrExplicit instruction, translate it back like llvm-spirv -r --spirv-target-env="CL2.0"

@MrSidims I've added above LIT and use opt to run.

**Comment by MrSidims** [comment]
Restarting CI with a new HEAD

**Review by MrSidims — APPROVED** [review]

File: test/transcoding/call-attribute.ll
Comment (svenvh) [inline test/transcoding/call-attribute.ll:1]: This isn't really a transcoding test (i.e., doesn't go LLVM IR -> SPIR-V -> LLVM IR), so I'd suggest moving it to a different place (outside of `transcoding/` at least, top level would be fine with me for now).

Comment (wenju-he) [reply test/transcoding/call-attribute.ll:1]: done, moved to top level

**Review by svenvh — APPROVED** [review]

**Comment by wenju-he** [comment]
@MrSidims could you please help to merge, thank you!

**Pull request merged**